### ISSUE TITLE
Cancel pending filters on new ones

### ISSUE THREAD

**Issue by splitbrain** [mroderick/plete#5]
When you autocomplete from a slow remote, your users might type faster than the remote can answer. It might even happen that multiple answers come back in the wrong order. A call to `filter()` should (maybe optionally) cancel all still pending filter operations.

**Comment by Lewiscowles1986** [comment]
not sure if it's against the browser compatibility goals, but I wouldn't mind taking a stab at this using https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort if @mroderick thinks it's worth it.

**Comment by mroderick** [comment]
I agree that it would be very nice to ensure that it is always the last request that gets displayed.

The `dataSrc` property is the (async) function provided by the developer.

This part of the API is intentionally generic, to allow the developer to use many different data sources, such as `fetch`, `xhr`, sockets, IndexedDB, etc.

One solution to the challenge of only showing the result of the last request, would be to allow the `dataSrc` function to return an array, consisting of a Promise and an `abort` function to call when making a new request.

Example from [How you can abort Fetch() request on a flight…
](https://itnext.io/how-you-can-abort-fetch-request-on-a-flight-830a639b9b92)

```js
const getGithubUser = (userName) => {
  const controller = new AbortController();
  const { signal } = controller; // get signal
  const params = {
    method: 'GET',
    signal, // bind controller to this fetch request using signal
  };
  
  const promise = fetch(`https://api.github.com/users/${userName}`, params)
    .then((data) => data.json());
    
  return [
    promise,
    controller.abort.bind(controller), // use bind to save controller context
  ];
};
```

It wouldn't be that challenging to add support for this approach to `plete`.

The change required would be to to update the `suggest` function at https://github.com/mroderick/plete/blob/0d03c670f443a1cc80bfb990201f764585fe4ebf/lib/suggest.js#L24-L25 so it detects if the returned value from `dataSrc` is a promise, or an array containing a promise and an `abort` function (otherwise throw an Error).

In the case of an array, a reference to the  `abort` function would be stored in the `state` of `plete`. When the user makes the next request, the `abort` function would then be called as the first thing.

This approach would still allow for many different transports to be used. For example, [you could use `axios`, which supports cancellation](https://github.com/axios/axios#cancellation).

@Lewiscowles1986 if you want to try this in a PR, I'd suggest the following:

- [ ] make a `fetch` and `AbortController` example in `index.html`, which you can use for developing the feature
- [ ] update `suggest.js` to detect array from `dataSrc` function
- [ ] store the `abort` function in `state`
- [ ] call `abort` before making new calls to `filter`

**Comment by mroderick** [comment]
I took a stab at this with @hexeberlin today, but ran into challenges in testing it.

We wanted to write some unit tests for the `suggest` function, but ran into issues with stubbing out ES modules when using `esm`. I've reached out to people for extra knowledge. Will update here as I learn more.

**Comment by mroderick** [comment]
This has been implemented in `plete@0.3.0`. Thank you for suggesting it! 🙇 

I'll update the documentation soon and improve the examples.
